### PR TITLE
[SPARK-43757][CONNECT] Change client compatibility from allow list to deny list

### PR DESCRIPTION
**Pull request by zhenlineo**
### What changes were proposed in this pull request?
Expand the client compatibility check to include all sql APIs.

### Why are the changes needed?
Enhance the API compatibility coverage

### Does this PR introduce _any_ user-facing change?
No, except it fixes a few wrong types and hides a few helper methods internally.

### How was this patch tested?
Existing tests.

### CODE REVIEW

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala [hunk 1]
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.util.Utils
 /**
  * Helper class for conversions between [[DataType]] and [[proto.DataType]].
  */
-object DataTypeProtoConverter {
+private[sql] object DataTypeProtoConverter {
```
Comment (zhenlineo) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala:29]: @grundprinzip Should the converters in common marked as public? Also see `LiteralValueProtoConverter` and `StorageLevelProtoConverter`

Can we modify them to be package protected?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala [hunk 2]
```diff
@@ -48,7 +48,7 @@ import org.apache.spark.util.{ThreadUtils, Utils}
  *   An unique identifier of the session which the artifact manager belongs to.
  * @param channel
  */
-class ArtifactManager(
+private[sql] class ArtifactManager(
```
Comment (zhenlineo) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/ArtifactManager.scala:51]: @vicennial Should the artifact and UDF related classes default to be public or package protected?

**Comment by zhenlineo** [comment]
cc @LuciferYang Could you have a look at this PR? Now the scala client contains enough methods that we can change to a deny list check.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 3]
```diff
@@ -38,7 +38,7 @@ import org.apache.spark.sql.streaming.{GroupState, GroupStateTimeout, OutputMode
  *
  * @since 3.5.0
  */
-abstract class KeyValueGroupedDataset[K, V] private[sql] () extends Serializable {
+class KeyValueGroupedDataset[K, V] private[sql] () extends Serializable {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:41]: Why is MiMa complaining about this?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 4]
```diff
@@ -122,65 +125,62 @@ object CheckConnectJvmClientCompatibility {
...
-      // Skip all shaded dependencies and proto files in the client.
-      ProblemFilters.exclude[Problem]("org.sparkproject.*"),
-      ProblemFilters.exclude[Problem]("org.apache.spark.connect.proto.*"),
+      // Skip unsupported packages
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.api.*"), // Java, Python, R
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.catalyst.*"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.columnar.*"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.connector.*"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.execution.*"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.expressions.*"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:139]: This is actually public API. What are we still missing here?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 5]
```diff
@@ -122,65 +125,62 @@ object CheckConnectJvmClientCompatibility {
...
+    val includedRules = Seq(IncludeByName("org.apache.spark.sql.*"))
     val excludeRules = Seq(
       // Filter unsupported rules:
       // Note when muting errors for a method, checks on all overloading methods are also muted.
 
-      // Skip all shaded dependencies and proto files in the client.
-      ProblemFilters.exclude[Problem]("org.sparkproject.*"),
-      ProblemFilters.exclude[Problem]("org.apache.spark.connect.proto.*"),
+      // Skip unsupported packages
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.api.*"), // Java, Python, R
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:134]: We need to tackle this at some point.

**Review by hvanhovell — APPROVED** [review]
LGTM

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 3 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:41]: due to `class org.apache.spark.sql.KeyValueGroupedDataset was concrete; is declared abstract in client version`

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala (hunk 4 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:139]: like `o.a.s.s.expressions.ReduceAggregator` and `o.a.s.s.expressions.MutableAggregationBuffer`, they do not need to exist in the connect-client module

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by hvanhovell** [comment]
Merging.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 6]
```diff
@@ -122,65 +125,62 @@ object CheckConnectJvmClientCompatibility {
...
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.COL_POS_KEY"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.DATASET_ID_KEY"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.curId"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.observe"),
+      ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.Observation"),
+      ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.Observation$"),
+      ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.ObservationListener"),
+      ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.ObservationListener$"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.queryExecution"),
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.encoder"),
```
[the remark this answers is not in the document]
Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:185]: yes. we changed the type from Encoder to AgnosticEncoder

**Pull request closed without merging**